### PR TITLE
[Core]: `aaz`: Support managed identity

### PR DESCRIPTION
**Pull request by necusjz**
**Related command**

**Description**

generate commands to support identity according to https://github.com/Azure/azure-cli/blob/20ffb2d35d59a104641d75aeb76b6f31885e26aa/doc/managed_identity_command_guideline.md

main command:
image

subcommand:
image

calculation logic be covered by `test_aaz_identity.py`

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py [hunk 1]
```diff
@@ -444,3 +445,113 @@ def to_serialized_data(self, processor=None, keep_undefined_in_list=False,  # py
         if processor:
             result = processor(self._schema, result)
         return result
+
+
+class AAZIdentityObject(AAZObject):
+    M_USER_ASSIGNED = "miUserAssigned"
+    S_USER_ASSIGNED = "userAssigned"
+    M_SYSTEM_ASSIGNED = "miSystemAssigned"
+    S_SYSTEM_ASSIGNED = "systemAssigned"
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:454]: Please move those const values out of the class. https://github.com/Azure/azure-cli/blob/414043d8f12fcec26ab9a191db2b28caba460d17/src/azure-cli-core/azure/cli/core/aaz/_field_value.py#L86-L90

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py [hunk 2]
```diff
@@ -444,3 +445,113 @@ def to_serialized_data(self, processor=None, keep_undefined_in_list=False,  # py
...
+
+        if not result and self._is_patch:
+            result = AAZUndefined
+
+        if processor:
+            result = processor(self._schema, result)
+
+        return result
+
+    def build_identity(self, data, result):
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:501]:
```suggestion
    def _build_identity(self, data, result):
```

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py [hunk 3]
```diff
@@ -444,3 +445,113 @@ def to_serialized_data(self, processor=None, keep_undefined_in_list=False,  # py
...
+
+                    else:
+                        result["type"] = "UserAssigned"
+
+                elif system_assigned or has_system_identity:
+                    result["type"] = "SystemAssigned"
+
+        else:  # main command
+            mi_user_assigned = data.get(self.M_USER_ASSIGNED, None)
+            mi_system_assigned = data.get(self.M_SYSTEM_ASSIGNED, None)
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:548]: Will the update command support mi_user_assigned and mi_system_assigned?

Comment (kairu-ms) [reply src/azure-cli-core/azure/cli/core/aaz/_field_value.py:548]: If the update command supported, then the mi_user_assigned or mu_system_asssigned could be undefined and you should think about the values from get as well to determine the final type.

Comment (necusjz) [reply src/azure-cli-core/azure/cli/core/aaz/_field_value.py:548]: update command won't support that,
https://github.com/necusjz/aaz-dev-tools/blob/96f0f5aa2e415d94c67986d9bb595c77658a0eef/src/aaz_dev/command/model/configuration/_arg_builder.py#L196

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py [hunk 4]
```diff
@@ -444,3 +445,113 @@ def to_serialized_data(self, processor=None, keep_undefined_in_list=False,  # py
...
+    def build_identity(self, data, result):
+        if not data:
+            return result
+
+        if action := data.get("action", None):  # subcommand
+            identities = set(result.pop("userAssignedIdentities", {}).keys())
+            has_system_identity = "systemassigned" in result.pop("type", "").lower()
+
+            user_assigned = data.get(self.S_USER_ASSIGNED, None)
+            system_assigned = data.get(self.S_SYSTEM_ASSIGNED, None)
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:510]: Please don't distinguish with AAZUndefined and None. Here the None mean user not defined?

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py [hunk 5]
```diff
@@ -444,3 +445,113 @@ def to_serialized_data(self, processor=None, keep_undefined_in_list=False,  # py
...
+                    else:  # remove all
+                        identities = {}
+
+                if identities:
+                    result["userAssignedIdentities"] = {k: {} for k in identities}
+                    if system_assigned or not has_system_identity:
+                        result["type"] = "UserAssigned"
+
+                    else:
+                        result["type"] = "SystemAssigned, UserAssigned"
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:526]: Please check whether the space after `,` can always work? You can find some swagger definition without space, For example https://github.com/Azure/azure-rest-api-specs/blob/b72e0199fa3242d64b0b49f38e71586066a8c048/specification/synapse/resource-manager/Microsoft.Synapse/stable/2021-05-01/workspace.json#L1014-L1019

Comment (necusjz) [reply src/azure-cli-core/azure/cli/core/aaz/_field_value.py:526]: after testing, it seems `"SystemAssigned,UserAssigned` always work. will change `"SystemAssigned, UserAssigned"` to `"SystemAssigned,UserAssigned"`

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py (hunk 4 above)
Comment (necusjz) [reply src/azure-cli-core/azure/cli/core/aaz/_field_value.py:510]: from https://github.com/necusjz/azure-cli/blob/8d78a5aca0f04072bb365db168b0a1c0ff1a05f7/src/azure-cli-core/azure/cli/core/aaz/_field_value.py#L477-L480, None means the user doesn't input the four parameters.

if we restrict the value of subcommand parameters to null, will this still be a problem?

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py (hunk 1 above)
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:454]: You can remove the "miUserAssigned" and "miSystemAssigned" and only use "userAssigned" and "systemAssigned" for consistence

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py (hunk 3 above)
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:548]: If no action then use user_assigned and system_assigned here

File: src/azure-cli-core/azure/cli/core/aaz/_field_value.py [hunk 6]
```diff
@@ -444,3 +445,93 @@ def to_serialized_data(self, processor=None, keep_undefined_in_list=False,  # py
...
+                    v = self[name].to_serialized_data(processor=processor, **kwargs)
+                    if v == AAZUndefined:
+                        continue
+
+                    if field_schema._serialized_name:
+                        name = field_schema._serialized_name
+
+                    if name in {"userAssigned", "systemAssigned"}:
+                        calculate_data[name] = v
+                        calculate_data["action"] = field_schema._flags.get("action", None)  # no action in GET operation
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_field_value.py:474]: By moving the action in IdentityObject, you can use the following way to get the action for calculation:
```
```suggestion
                        calculate_data["action"] = self._schema._flags.get("action", None) # no action in GET operation
```

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**